docs/decisions/0009-tagging-administrators.rst

```diff
@@ -0,0 +1,37 @@
+9. Taxonomy administrators
+==========================
+
+Context
+-------
+
+Taxonomy Administrators have the right to create, edit, populate, and delete taxonomies available globably for a given instance, or for a specific organization.
+
+How should these users be identified and their access granted?
+
+Decision
+--------
+
+A modified version of "course creator" access will be used to identify Taxonomy Administrators in Studio (ref `get_organizations`_):
+
+#. Global staff and superusers can create/edit/populate/delete Taxonomies for the instance or for any org key.
+
+#. Users who can create courses for "any organization" access can create/edit/populate/delete Taxonomies for the instance or for any org key.
+
+#. Users who can create courses only for specific organizations can create/edit/populate/delete Taxonomies with only these org keys.
+
+
+Permission #1 requires no external access, so can be enforced by the ``openedx_tagging`` app.
+
+But because permissions #2 + #3 require access to the edx-platform CMS model `CourseCreator`_, this access can only be enforced in Studio, and so will live under `cms.djangoapps.tagging` along with the ``ContentTag`` class. Tagging MVP must work for libraries v1, v2 and courses created in Studio, and so tying these permissions to Studio is reasonable for the MVP.
```

Comment:
Per OEP-9, these three rules that you listed above can be defined using [rules](https://github.com/dfunckt/django-rules)... with #1 defined in `openedx_tagging` and #2 and #3 defined in Studio. I _think_ that what that gets us is that any application can then use the standard Django API to query things like `user.has_perm('openedx_tagging.edit_taxonomy', taxonomy)` and get the right result without needing to be aware of how those permissions are defined or even where they are defined (in Studio).

Reply:
That would be really cool @bradenmacdonald , but can django-rules handle chaining multiple rule definitions together?

Ref [rules.RuleSet](https://github.com/dfunckt/django-rules#instance-methods-1):

* `add_rule` raises a KeyError if another rule by that name already exists.
* `set_rule` sets the rule, regardless of whether one already exists

But I don't see how to combine them?

Reply:
@pomegranited Isn't [combining rules](https://github.com/dfunckt/django-rules#combining-predicates) what you mean? It supports combining rules with "and", "or", "not", etc.

Reply:
@bradenmacdonald Yes, we can combine rules logically and explicitly -- what I meant was, there's no magic provided by django or django-rules to chain these calls to `set_rule`.

It's totally doable, it just takes a little care. For example: 

1. `openedx_tagging` calls `set_rule('openedx_tagging.edit_taxonomy', can_edit_taxonomy)` where `can_edit_taxonomy` checks e.g `user.is_staff | user.is_superuser`. This is fine, `openedx_tagging` can stand alone.
2. `openedx.features.tagging.SystemTaxonomy` might want to call `set_rule('openedx_tagging.edit_taxonomy', can_edit_taxonomy)` too, where its `can_edit_taxonomy` disallows editing for all SystemTaxonomy no matter who you are.
3. `openedx.features.tagging.OrgTaxonomy` might want to call `set_rule('openedx_tagging.edit_taxonomy', can_edit_taxonomy)` too, where its `can_edit_taxonomy` checks `openedx_tagging.can_edit_taxonomy` and the user's rights against the OrgTaxonomy.org.

But they can't all pile on and call `set_rule`, because the most recently called `set_rule` wins.

So to make sure `user.has_perm('openedx_tagging.edit_taxonomy', taxonomy)` returns the right thing, the `openedx.features.tagging` rules must be aware of its different Taxonomy subclasses, and explicitly combine with the `openedx_tagging` rules where needed.

Reply:
@pomegranited Ah, I see. Yeah it's too bad there's not a nice API for combining those. maybe we can add a lightweight way to do that?

Reply:
@bradenmacdonald Sure, will see.. I'll start by trying to use consistent convention, and if it ends up being abstractable into something we can reuse, that would be awesome :) Thanks for pointing me to OEP-9 here, and clarifying the end goal of these permission rules.